[PREVIEW] 🔝🥔

eqvinox opened the pull request:
This is purely a PREVIEW for any curious parties.

If you want to try running it, please refer to `tests/topotato/README` — YMMV, there are probably corner cases / incompatibilities that aren't handled correctly just yet.

qlyoung commented:
a preview of what

donaldsharp commented:
YOU SAY TOPOTATO, I SAY TOPOTATO

eqvinox commented:
> a preview of what

@qlyoung my attempt at replacing topotests with something that is less horrible. Namely, a potato.

Jokes aside, it's a new (blank-slate) test framework for FRR. It's been work in progress for a super long time (2019…), but at some point you gotta stop bikeshedding and start pushing.

qlyoung commented:
how's this interact with micronet?

gpnaveen commented:
@eqvinox - It will be better if we document what is the goal and what problem of topotest we are trying to solve. Please provide more info. Currently readme just explains how to run the tests.

eqvinox commented:
@qlyoung this has no interactions with micronet, there is no shared code between this and topotests. Feature-wise, this is a superset of micronet (it uses the same container setup, plus pidns, plus FreeBSD support via jails.) History-wise, this predates micronet, but I had made the determination that just removing ("fixing") topotests by removing mininet is not enough (which is what led me to build this on a blank slate.)

@gpnaveen this is a long-term replacement for topotests. I'm not joking when I'm arguing a potato is better than the current topotests system.

Chromico committed:
topotato: bgp default originate route map match

Authored-by: Chromico Rek <atech5122@gmail.com>
Signed-off-by: David Lamparter <equinox@diac24.net>

Chromico committed:
topotato: test bgp dont capability negotiate

Authored-by: Chromico Rek <atech5122@gmail.com>
Signed-off-by: David Lamparter <equinox@diac24.net>

eznix86 committed:
topotato: added doc for exabgp installation

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>

eznix86 committed:
topotato: added exabgp implementation

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>

eznix86 committed:
topotato: created demo for exabgp

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>

eznix86 committed:
topotato: fix name

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>

eznix86 committed:
topotato: Added conf validation

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>

Chromico committed:
topotato: test bgp set aspath

Authored-by: Chromico Rek <atech5122@gmail.com>
Signed-off-by: David Lamparter <equinox@diac24.net>

Chromico committed:
topotato: test bgp ebgp common subnet nexthop

Authored-by: Chromico Rek <atech5122@gmail.com>
Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: begin reworking javascript

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: log msgs in pcap-ng need a frame number

... they're "frames", so they need to be counted.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: stick more IDs in topology SVG

(to make things accessible from JavaScript)

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: python cleanup / polish

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
topotato: add some tempfile utilities

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: use python for inner-ns setup

Also create a locked file so the PID can be found.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: some rework/replumbing, add .interactive

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: start removing Instance class

pytest 7.x no longer has Instance classes

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: remove TopotatoInstance for pytest 7.x

pytest 7.x does not support python.Instance anymore.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: move makeitem hook to TopotatoItem

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: allow pytest 7 in requirements.txt

Should now work correctly.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
topotato: rework timing parameters

... to make relative timing work in the next step.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: properly use function-relative timing

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix scapy layer weirdness

sometimes we get Raw() instead of Ether(), not sure where/when/how that
broke...

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato/protomato: decode capture comments

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato/protomato: improve MLD display

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: return ItemGroup from iterator

... for future use in cross-referencing items.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix timing oddness in MLD basic test

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: clean up utils module

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: mark duplicates for pylint

Some of the Linux & FreeBSD code is bound to be similar, make pylint not
complain about that.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: begin cleaning up .pytestintegration

topotato.pytestintegration is the original hacking and dumping ground
for everything that didn't have a good place to go.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: stick black on .pytestintegration

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: remove python packet-to-HTML code

This is now exclusively in JavaScript.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: clean up .pretty

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix small mypy mistake in .multicast

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: stick black on .pretty

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
topotato: update documentation

TimedElement is not in .base anymore.  Also pytest docstrings leak in
for TopotatoItem, causing an undefined reference to "non-python tests".

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: type-hint and document .livelog

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: skip env setup for --collect-only

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
topotato: remove deprecated iter_logs()

Logs are no longer read from a file.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: update README with TODOs/known issues

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
topotato: create a .v1 to manage wildcard import

This fixes 2 bugs with one stone, er, python module - "topotato.*" no
longer causes a large chain import, and we get a versioning mechanism.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

eqvinox committed:
Merge branch 'master' into topotato

eqvinox committed:
topotato: add getlockpid tool

This is used to attach to a running topotato instance across pid
namespaces;  retrieving the PID of a locked file translates correctly
across the pidns boundary.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: use python3 for mcast-{rx,tx} tools

These will hopefully be removed soon anyway, but for the time being...

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: don't internal-error on startup failure

`instance.started_ts` won't exist if startup failed.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: improve startup failure behavior

- better report daemon start failures
- actually skip remainder of test

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix interactive halt w/o codeloc

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: don't break shutdown after startup fail

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: drop pytest-html patch

topotato is no longer using pytest-html

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

eqvinox committed:
topotato: extend protomato.js (BGP, PIM, IGMP)

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: re-add mac->name lookup in JS

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: change default to in-tree FRR location

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: remove unused topotato.css

None of the CSS definitions in this file have any effect.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: load config through vtysh for pim6d

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: strip some whitespace from vtysh

... this is otherwise extremely annoying to identify as a problem.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: update PIM6 basic test

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: lint cleanup pass

Fix a bunch of mypy/pylint/black issues that crept in.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

eqvinox committed:
topotato: add @topotatofunc(include_startup=)

... so we can add assertions for log messages printed at startup.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: bring mld_basic test up to date

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix scapy multi-packet send

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: rework pim_basic1 with scapy & mcast

Use scapy integration to send & receive packets, and MulticastReceiver
to join group.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

eqvinox committed:
topotato: pim_basic1 polish

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: convert pim_basic_topo2

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

hyperlinke committed:
topotato: added test for bgp comm-list delete

Signed-off-by: Noah Krishnamoorty <noah171202@gmail.com>

hyperlinke committed:
topotato: added test for bgp comm-list delete

Signed-off-by: Noah Krishnamoorty <noah171202@gmail.com>
(cherry picked from commit 6da05185cd2404b82b4b998f80f4fc7532ea8324)

Prerendered committed:
topotato: added test for bgp distance change

Signed-off-by: Brian Jugroo <jesjugroo@gmail.com>

Prerendered committed:
topotato: added test for bgp distance change

Signed-off-by: Brian Jugroo <jesjugroo@gmail.com>
(cherry picked from commit 9d03edfcdaf2e4fdcc04ec35dd94e3cedd52caad)

eqvinox committed:
Merge pull request #65 from N-011/topotato

eqvinox committed:
Merge pull request #66 from Prerendered/topotato

eqvinox committed:
topotato: fix bgp comm-list delete test

This wasn't quite working yet.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

eznix86 committed:
topotato: fix typing

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>
(cherry picked from commit 311baa6cb2cc7e723b3bf10b4bf52ae97eb824d3)

eznix86 committed:
topotato: fix typing

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>

eqvinox committed:
Merge pull request #67 from theArtechnology/fix-typing

eqvinox committed:
topotato: add .utils.Forked

Used for potatool.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: add .watch for potatool

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: WIP: potatool

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: don't make state files executable

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: drop setLevel calls on logger

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: more potatool/stop-pause features

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: include configs in HTML report

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: strip log messages from PDML in HTML

This significantly reduces the size of HTML reports.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: use better node IDs for JUnit XML compat

JUnit XML output splits the Node ID on `[` for parameters first, and
then `::` to build a class path.  Try to fit into that scheme.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

eqvinox committed:
topotato: potatool mostly-usable (still WIP)

Whole bunch more commands, random fixes & cleanup.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: eliminate some boilerplate

config/instance fixtures don't do much at this point.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: random style/lint/typing fixes

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: add neighbor iterator

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: IPv6 PIM prune propagation test

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge remote-tracking branch 'frr/master' into topotato

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

eqvinox committed:
topotato: add static route auto-calculation

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: render docstrings with docutils

Signed-off-by: David Lamparter <equinox@diac24.net>

MrTanoshii committed:
topotato: add ReconfigureFRR

Signed-off-by: Tsen Chee Vincent LEUNG YIN KO <vincent.leungyinko@gmail.com>
Co-authored-by: Bruno Bernard <contact.brunobernard@gmail.com>

MrTanoshii committed:
topotato: update test to use ReconfigureFRR

Signed-off-by: Tsen Chee Vincent LEUNG YIN KO <vincent.leungyinko@gmail.com>
Co-authored-by: Bruno Bernard <contact.brunobernard@gmail.com>

eznix86 committed:
topotato: fix missing dependency

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>

eznix86 committed:
topotato: added missing pyinotify

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>

Chromico committed:
topotato: migrate test_bgp_as_wide_bgp_identifier.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>

Chromico committed:
topotato: refactor test_bgp_as_wide_bgp_identifier.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>

Chromico committed:
topotato: refactor test_bgp_minimum_holdtime.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>
[DL: squashed bugfix commit]
Signed-off-by: David Lamparter <equinox@diac24.net>

Chromico committed:
topotato: refactor test_bgp_route_map_match_ipv6_nexthop.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>

Chromico committed:
topotato: refactor test_bgp_rmap_extcommunity_none.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>

eqvinox committed:
potatool: fix python 3.9 compat

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: help debug instance setup issues

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: make instancefn a class-scope fixture

<TestClass>.instancefn being set to the fixture itself is causing issues
in pytest where it thinks there's another fixture with the same name.
Start an attempt at detangling that...

Signed-off-by: David Lamparter <equinox@diac24.net>

Chromico committed:
topotato: test_bgp_default_originate_route_map_match_set.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>

Chromico committed:
topotato: test_bgp_aggregate_address_route_map.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>

Chromico committed:
topotato: test_bgp_aggregate_address_origin.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>

eqvinox committed:
topotato: allow overriding py ver in run_userns

Useful to run topotato on a specific python version rather than whatever
the default "python3" is.

Signed-off-by: David Lamparter <equinox@diac24.net>

Chromico committed:
topotato: improve test_bgp_set_aspath_replace.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>
Co-authored-by: Bruno Bernard <contact.brunobernard@gmail.com>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

Chromico committed:
topotato: test_bgp_sender_as_path_loop_detection.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>
Co-authored-by: Bruno Bernard <contact.brunobernard@gmail.com>

eqvinox committed:
topotato: fix mypy _ifix_name

Forgot to add annotations for this when fixing python <3.10

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix line endings

No \r\n please...

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix a bunch of formatting oddness

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: remove stray debug print

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix replaces for bgp aggregate rmap

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: make all topotatofunc args keywords

Removes the tripwire of getting the topology into the first router...

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: introduce delay to pim6_prune_propagate

Kernel MR stats are only fetched after 30 seconds...

Signed-off-by: David Lamparter <equinox@diac24.net>

eznix86 committed:
added exabgp to requirements

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>

eznix86 committed:
topotato: fix env on cli

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>

eznix86 committed:
topotato: refactor demo with new syntax

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>

Chromico committed:
topotato: test_disable_addpath_rx.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>
Co-authored-by: Bruno Bernard <contact.brunobernard@gmail.com>

eqvinox committed:
topotato: remove pytest-html / py.xml remnants

This clashes on python 3.11.  Properly shove vtysh output through
TimedElement/Timeline handling & JSON/Javascript.  This also fixes
timestamps being wrong for vtysh output.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix CLI output show/hide in JS

This got broken along the way (or was never quite complete?)

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix hiding all logs in JS

The logic for hiding all log messages wasn't quite correct.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix leftover old-style hook mark

... to get rid of a deprecation warning.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix more mkdir race conditions

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge pull request #82 from theArtechnology/topotato-exabgp-v2

eqvinox committed:
topotato: basic OSPFv3 PtMP test

Brings up ospf6d with `ipv6 ospf6 network point-to-multipoint` and
throws a little state change at it.

Does not work with current FRR master, needs `ospfv3-p2mp` branch.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: don't modify input lists in json cmp

The JSONCompareDirective items were being removed from the input list,
and then missing from 2nd compare onwards (for vtysh retries)

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: `_` to `-` in hostname

`_` is not valid in hostnames, `-` is not valid in variable names...
match made in heaven!

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: re-add OSPF short decode in HTML

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: re-add logarg marking in JS

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: pass data & restore xrefs lookups in JS

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: improve a bunch of typing hints

mypy has a (info-level) warning about type annotations in untyped
functions...

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: minor fixes for exabgp

_cmd should not be an attribute of the parent/container object, it'll
get overwritten by later Execute calls otherwise.  Also fix a bunch of
lint/typing issues.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: make JS resilient against missing xrefs

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: allow collapsing JSON nodes in HTML

E.g. `show ipv6 ospf database json` is rather long, make it a bit easier
to interact with.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: make some xrefs always monospaced

Some log messages are ASCII table-like.  Add a little JS usability.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: add collapse-children for JSON

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: rework vtysh methods, shutdown daemons

Replace `vtysh_fast` with `vtysh_polled`;  grab the PID from the vty
socket and use it to send SIGTERM to daemons when shutting down.  And
then wait until the daemons close their log output (indicating they have
exited.)

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: enable `debug memstats-at-exit`

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: record when log connection closes

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: show vtysh return codes & hilight errors

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato/vendor: import LCOV genhtml script

There's a few minor changes to make to this script.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato/vendor: record LCOV imported version

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato/vendor: better LCOV line number anchors

aka: github compatible line number anchors.  (Github uses #L1234.)

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: use pytest stash for interactive_session

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: isolate Linux/FreeBSD dispatch

... into a separate `.osdep` submodule.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: refactor out ClassHooks, move pytest foo

Replace `ClassHooks` with a `pytest_topotato_envcheck` hook.

Also move a whole bunch of stuff out of `topotato.pytestintegration` to
the places it's actually used/relevant.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: don't burst missed timing ticks

When setting up a retry, the timeout is relative to the beginning of the
test generator function.  If previous test items accumulated some delay,
don't burst-retry attempts that we can no longer do because they lie in
the past.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato/vendor: support .xz compression in LCOV genhtml

... so the coverage data in topotato reports can be compressed.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: use AutoFixture everywhere

Signed-off-by: David Lamparter <equinox@diac24.net>

Chromico committed:
topotato: test_bgp_aspath_zero.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>
Co-authored-by: Bruno Bernard <contact.brunobernard@gmail.com>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge remote-tracking branch 'frr/master' into topotato

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

eqvinox committed:
topotato: undo pytest.stash changes

... it's not supported in pytest 6 :(

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

eqvinox committed:
topotato: stick `debug mld` into PIM tests

... so it actually gets the log messages it expects.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: show TCP flags in JS render

Really useful to see SYN/FIN/RST.  (ACK, not so much...)

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: optionally use gevent

generating the coverage reports takes quite a while and can run in the
background while we keep running other tests.  gevent makes this
painless through the use of Greenlets.

Signed-off-by: David Lamparter <equinox@diac24.net>

eznix86 committed:
topotato: fix gitignore

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>

eqvinox committed:
topotato/vendor: sort by file extension in LCOV

Rather have the .h files sorted below the .c files, their coverage
percentages are kinda bogus.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: add finish() pass to pretty

... to be used for generating an index.html and a summary coverage
report.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: random black formatting adjustments

Apparently I updated to a newer version of black which likes its empty
lines with extra flavor.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: cleanup & improve vtysh output handling

Don't eat up the error message for unsuccessful CLI calls.  Also remove
TopotatoItem.commands, it's no longer used.  And flag .match_for, even
though it's not used yet.

Signed-off-by: David Lamparter <equinox@diac24.net>

Chromico committed:
topotato: test_bgp_ebgp_common_subnet_nexthop_unchanged.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>
Co-authored-by: Bruno Bernard <contact.brunobernard@gmail.com>

Chromico committed:
topotato: test_bgp_extended_optional_parameters_length.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>

eqvinox committed:
topotato: add `_bgp` into addpath test filename

... for consistency.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

eqvinox committed:
topotato: fix debugs for pim6_prune_propagate

pim6d no longer has all debugs enabled unconditionally.

Reduce zebra debugs while at it, the report is excessively large.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: repair & clean up `--run-topology`

Due to some pytest changes, fixtures can no longer be explicitly
accessed from the collect function.  So for `--run-topology`, just let
the startup item execute normally and stop after that.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix pylint complaints in .toponom

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: improve auto-numbering sort order

Sort systems named "dut", "r" or "rt" before others.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: allow overriding auto-numbering

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: remove outdated bits from docs

config_fixture / instance_fixture are on their way out.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: random docs whitespace

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: set docs language to English

... mostly to shut up the warning from sphinx-build.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: extend docs introduction section

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: remove {config,instance}_fixture

Now handled by AutoFixture, in preparation for some refactoring.

Signed-off-by: David Lamparter <equinox@diac24.net>

Chromico committed:
topotato: test_bgp_set_local_preference_add_subtract

Signed-off-by: Nathan Mangar <nathan@thundergear.io>

eznix86 committed:
topotato: add vagrant setup

- Ignore vagrant config
- Added utility commands
- Added Vagrant and configs and docs
- remove unnecessary vagrant config
- restore missing make command
- Add comments and extra code for Vagrant
- Add missing script config
- fix typo
[DL: whitespace & squashed]

Signed-off-by: Bruno Bernard <contact.brunobernard@gmail.com>
Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: add .generatorwrap, catch missing yield

Forgetting the "yield from" in a topotatofunc is... very annoying and
very hard to catch.  Add infrastructure to try catching this.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: use .generatorwrap in TestBase

Now forgetting the "yield from" on a .make() call results in an
exception.

Signed-off-by: David Lamparter <equinox@diac24.net>

gromit1811 committed:
topotato: add inversion option to AssertPacket

... to check that a specific packet did *not* show up.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: ignore args with default in AssertPacket

Allow using default values to capture values into the definition of the
packet matching function.  Directly referring to a variable from the
outer scope otherwise has the problem that the variable may change
meanwhile, e.g.

    for i in range(0, 8):
        def foo():
            return i
        something(foo)

When something holds on to foo, even though there are 8 copies of foo,
all 8 copies will return 7 since i is updated.  The "fix" is:

    for i in range(0, 8):
        def foo(i=i):
            return i
        something(foo)

Which will have foo return 0...7 as expected.

Reported-by: Martin Buck <mb-tmp-tvguho.pbz@gromit.dyndns.org>
Signed-off-by: David Lamparter <equinox@diac24.net>

Chromico committed:
topotato: migrate test_bgp_default_originate.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>

eqvinox committed:
topotato: get error locations in jinja templates

Tell jinja2 what file (and line) templates originated from, so
exceptions/syntax errors can be reported appropriately.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: mark staticd integrated-only

With `mgmtd` on master, `staticd` no longer accepts an `-f` option to
load a config file.  Put it on integrated-only (which will work on older
versions too.)

At some point everything should be integrated-only anyway...

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: improve SKIPPED reporting

Keep a little bit of information so the SKIPPED line printed is more
useful.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix AssertPacket nits

mypy/pylint/black are complaining.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: add .pylintrc

Disable some stuff that just doesn't matter (for now.)

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: random black formatting fix

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix template whitespace

whitespace-only change to get consistent formatting *inside* the jinja2
config template strings.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge branch 'master' into topotato

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

Chromico committed:
topotato: fix type error generatorwrap.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>

eqvinox committed:
topotato: move FRR into its own submodule

Make a bit more clear what is FRR specific code.  Also easier to see
where FRR is currently hardwired into topotato.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: extricate FRRNetworkInstance

FRRNetworkInstance has really become counterproductive in a way;
there's nothing about a test network as a whole that makes it "FRR".
Some virtual routers inside of it are probably FRR, some might be hosts,
some might be exabgp...

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: decouple FRRSetup from FRRConfigs

FRRConfigs was also covering the FRR install encapsulation on its class
level, which is a bit of an unnecessary entanglement.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: split off frr.templating

move TemplateUtils into .templating submodule.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: add "vtysh" target for AssertVtysh

Allow fork()+exec() of actual vtysh, to run commands that aren't
directly available in-daemon.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix random leftover style/lint nits

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: add pre-commit hooks

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: improve Network/Router abstraction

Introduce abstract base classes for NetworkInstance & virtual routers,
and move some pieces into better places behind abstraction.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix test_rip.py

TODO: find a better way to deal with whitespace...

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: add mgmtd support

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: improve DaemonCrash exception reporting

The "short" exception summary was just plain empty - not very helpful.

Signed-off-by: David Lamparter <equinox@diac24.net>

Chromico committed:
topotato: test_bgp_max_med_on_startup.py

Signed-off-by: Nathan Mangar <nathan@thundergear.io>

eqvinox committed:
topotato: job control for potatool

Make ^C and ^Z work and not wreck the terminal.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: don't fail with typing.{Self,TypeAlias}

These imports are only needed for type checking and don't exist on
Python 3.8;  move the import behind TYPE_CHECKING guards.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: rework/fix deindent()

It wasn't quite working as expected if the first line had more indent
than further lines.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix potatool CLI decorator

... it was "voiding out" @staticmethod/@classmethod.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: allow empty backtrace (for interactive)

Instantiating a test item from an interactive or pdb shell won't have
any useful traceback - avoid failing with an exception in that case.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: add _yield_from to interactive pause

Called as e.g. `_yield_from(AssertVtysh.make(...))`, this can be used to
interactively try and poke at test items.  It's not 100% identical
(completely bypasses pytest collection) but a significant DX improvement
regardless.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: fix docs after FRR restructuring

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: document pause-on-fail and potatool

Extremely important tools.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: allow instantiating off TopotatoNetwork

Preparation to get per-host roles working, not useful on its own just
yet.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: deprecation warn on FRRRouterNS.configs

Working towards making this private.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: update install instructions

Installing system packages with apt-get is recommended.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
topotato: remove duplicated file

This got mixed up in merging PRs somewhere.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge remote-tracking branch 'frr/master' into topotato

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

eqvinox committed:
topotato: don't depend on `deprecated` package

Really only want this for development/topotato hacking.

Signed-off-by: David Lamparter <equinox@diac24.net>

eqvinox committed:
Merge branch 'master' into upstreaming

eqvinox committed:
Merge remote-tracking branch 'topotato/upstreaming' into topotato

riw777 commented:
@frrbot autoclose in 1 month
